src/helper/ecdsa.rs

```diff
@@ -0,0 +1,223 @@
+use crate::{
+    bail,
+    crypto::{hash::HashAlgorithm, public_key::PublicKeyAlgorithm},
+    errors::Result,
+    packet::PublicKey,
+    types::{
+        EcdsaPublicParams, EskType, Fingerprint, KeyId, KeyVersion, Mpi, PkeskBytes,
+        PublicKeyTrait, PublicParams, SecretKeyTrait, SignatureBytes, Version,
+    },
+};
+
+impl<C> PgpPublicKey for ecdsa::VerifyingKey<C>
+where
+    Self: PgpEcdsaPublicKey,
+    C: PrimeCurve + CurveArithmetic,
+{
+    const PGP_ALGORITHM: PublicKeyAlgorithm = PublicKeyAlgorithm::ECDSA;
+
+    fn pgp_parameters(&self) -> PublicParams {
+        let key = self.ecdsa_public_key();
+        PublicParams::ECDSA(key)
+    }
+}
+
+/// Encoding for an ecdsa public key
+pub trait PgpEcdsaPublicKey {
+    /// public key parameter for a given public key
+    fn ecdsa_public_key(&self) -> EcdsaPublicParams;
+}
+
+impl PgpEcdsaPublicKey for p256::ecdsa::VerifyingKey {
+    fn ecdsa_public_key(&self) -> EcdsaPublicParams {
+        let key = self.into();
+        let p = Mpi::from_raw(self.to_sec1_bytes().to_vec());
+        EcdsaPublicParams::P256 { key, p }
+    }
+}
+
+/// [`signature::Signer`] backed signer for PGP.
+pub struct EcdsaSigner<T, C> {
+    inner: T,
+    public_key: PublicKey,
+    _signature: PhantomData<C>,
+}
+
+impl<C, T> EcdsaSigner<T, C>
+where
+    T: Keypair,
+    T::VerifyingKey: PgpPublicKey,
+{
+    /// Create a new signer with a given public key
+    pub fn new(inner: T, created_at: DateTime<Utc>) -> Result<Self> {
+        let public_key = PublicKey::new(
+            Version::New,
+            KeyVersion::V4,
+            <T as Keypair>::VerifyingKey::PGP_ALGORITHM,
+            created_at,
+            None,
+            inner.verifying_key().pgp_parameters(),
+        )?;
+
+        Ok(Self {
+            inner,
+            public_key,
+            _signature: PhantomData,
+        })
+    }
+}
+
+impl<C, T> Keypair for EcdsaSigner<T, C>
+where
+    T: Keypair,
+{
+    type VerifyingKey = T::VerifyingKey;
+
+    fn verifying_key(&self) -> Self::VerifyingKey {
+        self.inner.verifying_key()
+    }
+}
+
+impl<C, T> EcdsaSigner<T, C>
+where
+    C: PrimeCurve + DigestPrimitive,
+    SignatureSize<C>: ArrayLength<u8>,
+    T: PrehashSigner<ecdsa::Signature<C>>,
+    C::Digest: PgpHash,
+{
+    fn sign_prehash(&self, hash: HashAlgorithm, prehash: &[u8]) -> Result<Vec<Vec<u8>>> {
+        if C::Digest::HASH_ALGORITHM != hash {
+            bail!(
+                "Signer only support {expected:?}, found {found:?}",
+                expected = C::Digest::HASH_ALGORITHM,
+                found = hash
+            );
+        }
+        // TODO: check digest size
+        let signature = self.inner.sign_prehash(prehash)?;
+        let (r, s) = signature.split_bytes();
+        Ok(vec![r.to_vec(), s.to_vec()])
+    }
+}
+
+impl<C, T> fmt::Debug for EcdsaSigner<T, C> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("EcdsaSigner").finish()
```

Comment:
Just out of curiosity why not simply derive the Debug? Is this about hiding the `inner` which may contain private bits?

Reply:
inner then needs to `impl Debug` which might not be feasible yeah.
I'll add the `public_key`